### PR TITLE
754 libbeat panic on exit

### PR DESCRIPTION
**Pull request by cyrilleverrier**

### CODE REVIEW

File: filebeat/tests/system/filebeat.py [hunk 1]
```diff
@@ -254,6 +254,9 @@ def all_fields_are_expected(self, objs, expected_fields,
                     raise Exception("Unexpected key '{}' found"
                                     .format(key))
 
+    def did_not_panic(self, output_file="output/filebeat"):
+        return self.log_contains("panic: ") == False
```
Comment (ruflin) [inline filebeat/tests/system/filebeat.py]: :+1: Like this we prevent it from happening in the future.

File: filebeat/tests/system/test_crawler.py [hunk 2]
```diff
@@ -48,6 +48,9 @@ def test_fetched_lines(self):
         # Check that output file has the same number of lines as the log file
         assert iterations == len(output)
 
+        # Check that the application did not panic
+        assert self.did_not_panic()
+
```
Comment (tsg) [inline filebeat/tests/system/test_crawler.py]: I wonder if we can't puț these în the tearDown method to avoid duplication. I guess panics are never good.

Comment (cyrilleverrier) [inline filebeat/tests/system/test_crawler.py]: I thought that too. It's better to test for panics by default in the test cases.

File: libbeat/beat/beat.go [hunk 3]
```diff
@@ -218,7 +217,7 @@ func (b *Beat) Run() error {
 	// Callback is called if the processes is asked to stop.
 	// This needs to be called before the main loop is started so that
 	// it can register the signals that stop or query (on Windows) the loop.
-	service.HandleSignals(b.Exit)
+	service.HandleSignals(b.Stop)
```
Comment (ruflin) [inline libbeat/beat/beat.go]: Exit was put here on purpose as Stop could also be used to later use again Start. In the case of exit, the complete beat exits which I think should be the case when CTRL-C is used. If Exit is used, the b.exit channel is probably never closed. Let me see if we can find an alternative solution here.

File: filebeat/tests/system/test_crawler.py (hunk 2 above)
Comment (andrewkroh) [inline filebeat/tests/system/test_crawler.py]: I think checking the exit codes would also be good. If the process exits due to a panic the exit code should be non-zero.

File: libbeat/beat/beat.go (hunk 3 above)
Comment (ruflin) [inline libbeat/beat/beat.go]: An alternative here could be:

```

var callback sync.Once

// Exiting beat -> shutdown
func (b *Beat) Exit() {

	callback.Do(func() {
		logp.Info("Start exiting beat")
		close(b.exit)
	})
}
```

This makes sure Exit is only called once. It will not solve the problem that the filebeat and topbeat Stop function to not cleanup everything yet before stopping.

**Comment by ruflin** [comment]
@cyrilleverrier Here is a gist I created with a potential fix in beat.go: https://gist.github.com/ruflin/228e96bbd9c9e16bfdbe The additional change I made in addition to the above, is moving the cleanup to the Stup function (where I think it belongs). Combined with your spooler fix it should probably prevent the panic issue.

**Comment by cyrilleverrier** [comment]
@ruflin thank you for the fix. I built upon it to make improvements.
Also added more system tests for filebeat and mockbeat.

One source code incompatibility:
I prefer to let the main function Exit with the right status code rather than doing it in libbeat.
It allows the developer of beats to add additional logic in the main function if needed.

for instance, filebeat/main.go:
```
func main() {
	if err := beat.Run(Name, "", filebeat.New()); err != nil {
		os.Exit(1)
	}
}
```


If you prefer not, I can move the code back to libbeat

File: filebeat/crawler/crawler.go [hunk 4]
```diff
@@ -42,7 +42,10 @@ func (c *Crawler) Start(prospectorConfigs []config.ProspectorConfig, eventChan c
 		logp.Debug("prospector", "File Configs: %v", prospectorConfig.Paths)
 
 		prospector, err := NewProspector(prospectorConfig, c.Registrar, eventChan)
-		c.prospectors = append(c.prospectors, prospector)
+
+		if prospector != nil {
```
Comment (ruflin) [inline filebeat/crawler/crawler.go]: Instead of introducing the if clause, it would probably best to move this line after the error check.

**Comment by ruflin** [comment]
I quite like the proposal above. I didn't do it initially because it changes the libbeat API. But to be honest it is much nicer and gives more flexibility to the developer. The old impelmentations will still work, it will just not return 1 in case of an error.

**Comment by ruflin** [comment]
@cyrilleverrier Could you rebase on master just to make sure we have all the most recent changes in? I made some changes around the above code very recently.

File: filebeat/tests/system/test_crawler.py (hunk 2 above)
Comment (ruflin) [inline filebeat/tests/system/test_crawler.py]: Agree, we should definitively move both to tear down methods in TestCase. I'm sure there is also an annotation to skip TearDown for exceptions where for example exit should be 1 and not 0 (which is probably default)

**Comment by cyrilleverrier** [comment]
@ruflin : all beats are now updated: main.go 
system tests have been refactored
branch is rebased on the up-to-date master branch

File: filebeat/crawler/crawler.go [hunk 5]
```diff
@@ -41,11 +41,10 @@ func (c *Crawler) Start(prospectorConfigs []config.ProspectorConfig, eventChan c
...
 		logp.Debug("prospector", "File Configs: %v", prospectorConfig.Paths)
 
-		prospector, err := NewProspector(prospectorConfig, c.Registrar, eventChan)
-		c.prospectors = append(c.prospectors, prospector)
-
-		if err != nil {
-			return fmt.Errorf("Error in initing prospector: %s", err)
+		if prospector, err := NewProspector(prospectorConfig, c.Registrar, eventChan); err != nil {
+			return fmt.Errorf("Error in initing prospector: %v", err)
+		} else {
```
Comment (ruflin) [inline filebeat/crawler/crawler.go]: The else would not even be needed, as the if has a return.

**Comment by ruflin** [comment]
Thanks a lot. Your changes triggered me to make improvements on our system tests as there is way too much duplicated code. See:

* https://github.com/elastic/beats/pull/762
* https://github.com/elastic/beats/pull/763
* https://github.com/elastic/beats/pull/764

I'm now thinking if we should split up the changes in at least 2 parts: Python changes and the following Golang changes.

**Comment by ruflin** [comment]
@cyrilleverrier Perhaps it is easiest to briefly chat?

File: filebeat/config/config_test.go [hunk 6]
```diff
@@ -82,10 +83,15 @@ func TestGetConfigFiles_Dir(t *testing.T) {
 	files, err := getConfigFiles(absPath)
 
 	assert.Nil(t, err)
-	assert.Equal(t, 2, len(files))
+	assert.Condition(t, func() bool { return len(files) >= 2 })
```
Comment (ruflin) [inline filebeat/config/config_test.go]: Is that change somehow related to the other changes?

File: filebeat/main.go [hunk 7]
```diff
@@ -3,6 +3,7 @@ package main
 import (
 	filebeat "github.com/elastic/beats/filebeat/beat"
 	"github.com/elastic/beats/libbeat/beat"
+	"os"
```
Comment (ruflin) [inline filebeat/main.go]: We normally separate the internal and external exports. See https://github.com/elastic/beats/blob/master/filebeat/harvester/log.go#L7

File: filebeat/config/config_test.go (hunk 6 above)
Comment (cyrilleverrier) [inline filebeat/config/config_test.go]: Yes, I added `invalid.yml` file in the same directory, used by one of the system test.

I rewrote the unit tests to check for the existence of `config.yml` and `config2.yml`. But there might be more files.

File: filebeat/crawler/crawler.go (hunk 5 above)
Comment (cyrilleverrier) [inline filebeat/crawler/crawler.go]: agree

File: filebeat/main.go (hunk 7 above)
Comment (cyrilleverrier) [inline filebeat/main.go]: ok, can you fix it?

Comment (ruflin) [inline filebeat/main.go]: yes

File: filebeat/config/config_test.go (hunk 6 above)
Comment (ruflin) [inline filebeat/config/config_test.go]: ok

**Comment by ruflin** [comment]
The following things were done in this PR:
* [ ] Prevent beats when stopping from panic
* [ ] Introduce basic tests for all beats (config check, crash check)
* [ ] Check for exit code when beat stops in system tests
* [ ] Check for panic when system tests end

As discussed with @cyrilleverrier I will take the code of this PR and split it up into multiple smaller PR's.

File: libbeat/beat/beat.go [hunk 8]
```diff
@@ -248,27 +254,36 @@ func (b *Beat) Run() error {
...
-	return err
 }
 
-// Stop calls the beater Stop action.
-// It can happen that this function is called more then once.
-func (beat *Beat) Stop() {
-	logp.Info("Stopping Beat")
-	beat.BT.Stop()
-}
+var callback sync.Once
```
Comment (ruflin) [inline libbeat/beat/beat.go]: This could be move also into Beat to not have local variables.

**Comment by ruflin** [comment]
@cyrilleverrier Not sure if you saw the checklist I added above. I will keep it up-to-date for the progress.

**Comment by cyrilleverrier** [comment]
@ruflin thanks, looking forward to seeing the updates :+1:

**Comment by ruflin** [comment]
@cyrilleverrier I think all of the major issues are now covered with PR's, but not all are closed yet. There is some cleanup you did in this PR for example related to invalid config file checking. This is not done in an other PR yet. Feel free to open a new PR with these changes, but best is to wait until all the other related PR's were merged to prevent conflicts. I will close this as I think all further points should go in a new PR. 

Thanks a lot for preparing all the base work in the PR and pushing this forward.
